src/pseudopeople/interface.py

```diff
@@ -46,65 +46,89 @@ def _generate_form(
     source = Path(source) / form_file_name
     data_paths = [x for x in source.glob(f"{form_file_name}*")]
     if not data_paths:
-        logger.warning(
+        raise ValueError(
             f"No datasets found at directory {str(source)}. "
             "Please provide the path to the unmodified root data directory."
         )
-        return None
     suffix = set(x.suffix for x in data_paths)
     if len(suffix) > 1:
         raise TypeError(
             f"Only one type of file extension expected but more than one found: {suffix}. "
             "Please provide the path to the unmodified root data directory."
         )
     noised_form = []
-    columns_to_keep = [c for c in form.columns]
     for data_path in data_paths:
-        if data_path.suffix == ".hdf":
-            with pd.HDFStore(str(data_path), mode="r") as hdf_store:
-                data = hdf_store.select("data", where=year_filter["hdf"])
-        elif data_path.suffix == ".parquet":
-            data = pq.read_table(data_path, filters=year_filter["parquet"]).to_pandas()
-        else:
-            raise ValueError(
-                "Source path must either be a .hdf or a .parquet file. Provided "
-                f"{data_path.suffix}"
-            )
-        if not isinstance(data, pd.DataFrame):
-            raise TypeError(
-                f"File located at {data_path} must contain a pandas DataFrame. "
-                "Please provide the path to the unmodified root data directory."
-            )
-
-        # Coerce dtypes prior to noising to catch issues early as well as
-        # get most columns away from dtype 'category' and into 'object' (strings)
-        for col in columns_to_keep:
-            if col.dtype_name != data[col.name].dtype.name:
-                data[col.name] = data[col.name].astype(col.dtype_name)
+        data = _load_data_from_path(data_path, year_filter)
 
+        data = _coerce_dtypes(data, form)
+        data = _reformat_dates_for_noising(data, form)
         noised_data = noise_form(form, data, configuration_tree, seed)
-        noised_data = _extract_columns(columns_to_keep, noised_data)
+        noised_data = _extract_columns(form.columns, noised_data)
```

Comment:
I do like this refactor quite a bit. I used to be quite good doing this for the sole purpose of de-spaghetti-ing code but seem to have gotten sloppy.

Reply:
Everything's a tradeoff :) Here's the legendary John Carmack with a timely tweet on the subject https://twitter.com/ID_AA_Carmack/status/1646634954703421441